azkaban-exec-server/src/main/java/azkaban/execapp/FlowPreparer.java

```diff
@@ -84,328 +78,202 @@ synchronized void incrementCacheMiss() {
+      // conditions which could arise when multiple threads are downloading/deleting/hard-linking
+      // the same project. But it doesn't prevent multiple executor processes interfering with each
+      // other triggering race conditions. So it's important to operationally make sure that only
+      // one executor process is setting up flow execution against the shared project directory.
+      long criticalSectionStartTime = -1;
+      File execDir = null;
+
+      synchronized (this) {
+        criticalSectionStartTime = System.currentTimeMillis();
+        if (!project.getInstalledDir().exists()) {
+          // If new project is downloaded and project dir cache clean-up feature is enabled, then
+          // perform clean-up if size of all project dirs exceeds the cache size.
+          if (isDownloaded && this.projectCacheCleaner.isPresent()) {
+            this.projectCacheCleaner.get()
+                .deleteProjectDirsIfNecessary(project.getDirSizeInByte());
+          }
 
-      if (isProjectCacheSizeLimitEnabled() && !isFileCountEqual(projectVersion, linkCount)) {
-        throw new Exception(String.format("File count check failed for execid: %d, project dir %s"
-                + " are being deleted when setting this execution up",
-            flow.getExecutionId(), projectVersion.getInstalledDir()));
+          // Rename temp dir to a proper project directory name.
+          Files.move(tempDir.toPath(), project.getInstalledDir().toPath());
+        }
+        execDir = setupExecutionDir(project.getInstalledDir(), flow);
       }
 
-      log.info(String
-          .format("Flow Preparation complete. [execid: %d, path: %s]", flow.getExecutionId(),
-              execDir.getPath()));
-    } catch (final Exception e) {
-      log.error("Error in setting up project directory: " + this.projectsDir + ", Exception: " + e);
-      cleanup(execDir);
-      throw new RuntimeException(e);
+      final long flowPrepCompletionTime = System.currentTimeMillis();
+      log.info("Flow preparation completed in {} sec(s), out ot which {} sec(s) was spent inside "
+              + "critical section. [execid: {}, path: {}]",
+          (flowPrepCompletionTime - flowPrepStartTime) / 1000,
+          (flowPrepCompletionTime - criticalSectionStartTime) / 1000,
+          flow.getExecutionId(), execDir.getPath());
+    } catch (final Exception ex) {
+      log.error("Error in preparing flow execution {}", flow.getExecutionId(), ex);
+      throw new ExecutorManagerException(ex);
+    } finally {
+      if (projectFileHandler != null) {
+        projectFileHandler.deleteLocalFile();
+      }
+      FileIOUtils.deleteDirectory(tempDir);
     }
   }
 
-  private void cleanup(final File execDir) {
-    if (execDir != null) {
-      try {
-        FileUtils.deleteDirectory(execDir);
-      } catch (final IOException e) {
-        throw new RuntimeException(e);
-      }
+  private File setupExecutionDir(final File installedDir, final ExecutableFlow flow)
```

Comment:
i don't have a number yet, but my estimation would be 5-10 secs for a large project(~thousands files), since it takes several seconds for an UMP flow to pass over preparation stage if the cluster is not busy.